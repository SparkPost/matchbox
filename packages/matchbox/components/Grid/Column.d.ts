export default Column;
declare function Column({ children, ...props }: {
    [x: string]: any;
    children: any;
}): any;
declare namespace Column {
    const displayName: string;
    namespace propTypes {
        const xs: any;
        const sm: any;
        const md: any;
        const lg: any;
        const xl: any;
        const xsOffset: any;
        const smOffset: any;
        const mdOffset: any;
        const lgOffset: any;
        const children: any;
    }
}
