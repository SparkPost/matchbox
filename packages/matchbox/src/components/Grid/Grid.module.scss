@import '../../styles/config';

// TODO Create responsive helpers for these
$breakpoints: (
  sm: 450px,     // 30em,
  md: 720px,     // 48em,
  lg: 960px,     // 64em,
  xl: 1200px,    // 80em,
  // xxl: 1470px    // 98em,
);

// Gutter Width
$gutter-width: rem(18);

// Number of Columns
$grid-columns: 12;

// Time to get flexy ðŸ’ª
.Grid {
  box-sizing: border-box;
  display: flex;
  flex: 0 1 auto;
  flex-direction: row;
  flex-wrap: wrap;
  margin-left: -$gutter-width;
}

.start-xs {
  justify-content: flex-start;
  text-align: start;
}

.center-xs {
  justify-content: center;
  text-align: center;
}

.end-xs {
  justify-content: flex-end;
  text-align: end;
}

.top-xs {
  align-items: flex-start;
}

.middle-xs {
  align-items: center;
}

.bottom-xs {
  align-items: flex-end;
}

.around-xs {
  justify-content: space-around;
}

.between-xs {
  justify-content: space-between;
}

@each $bp, $bp-size in $breakpoints {
  @media only screen and (min-width: $bp-size) {
    .start-#{$bp} {
      justify-content: flex-start;
      text-align: start;
    }

    .center-#{$bp} {
      justify-content: center;
      text-align: center;
    }

    .end-#{$bp} {
      justify-content: flex-end;
      text-align: end;
    }

    .top-#{$bp} {
      align-items: flex-start;
    }

    .middle-#{$bp} {
      align-items: center;
    }

    .bottom-#{$bp} {
      align-items: flex-end;
    }

    .around-#{$bp} {
      justify-content: space-around;
    }

    .between-#{$bp} {
      justify-content: space-between;
    }
  }
}

@mixin flex-col {
  box-sizing: border-box;
  flex: 0 0 auto;
  padding-left: $gutter-width;
  // margin-bottom: $gutter-width;
  text-align: left;
}

// Column
.Column {
  @include flex-col;
  flex-grow: 1;
  flex-basis: 0;
  max-width: 100%;
}

@for $i from 1 through $grid-columns {
  $width: 100% / $grid-columns * $i;

  .xs-#{$i} {
    @include flex-col;
    flex-basis: $width;
    max-width: $width;
  }

  .xs-offset-#{$i} {
    margin-left: $width;
  }
}

@each $bp, $bp-size in $breakpoints {
  @media only screen and (min-width: $bp-size) {
    @for $i from 1 through $grid-columns {
      $width: 100% / $grid-columns * $i;

      .#{$bp}-#{$i} {
        @include flex-col;
        flex-basis: $width;
        max-width: $width;
      }

      .#{$bp}-offset-#{$i} {
        margin-left: $width;
      }
    }
  }
}

// TODO order
// .first-xs {
//   order: -1;
// }
//
// .last-xs {
//   order: 1;
// }
