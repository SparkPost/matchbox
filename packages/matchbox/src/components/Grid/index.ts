export { default as Grid } from './Grid';
export * from './Column';
