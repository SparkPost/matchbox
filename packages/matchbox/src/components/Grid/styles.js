import { tokens } from '@sparkpost/design-tokens';

const gutterWidth = tokens.spacing_400;
export const breakpoints = ['xs', 'sm', 'md', 'lg', 'xl'];
const alignmentStyleMap = {
  start: {
    justifyContent: 'flex-start',
    textAlign: 'start',
  },
  center: {
    justifyContent: 'center',
    textAlign: 'center',
  },
  end: {
    justifyContent: 'flex-end',
    textAlign: 'end',
  },
  top: {
    alignItems: 'flex-start',
  },
  middle: {
    alignItems: 'center',
  },
  bottom: {
    alignItems: 'flex-end',
  },
  around: {
    justifyContent: 'space-around',
  },
  between: {
    justifyContent: 'space-between',
  },
};

const wrapStyleWithMediaQuery = (breakpoint, style) => {
  const mediaQueryWidth = tokens[`mediaQuery_${breakpoint}`];

  // do not wrap if no style for breakpoint or breakpoint is extra small
  if (!Object.keys(style).length || breakpoint === 'xs') {
    return style;
  }

  return {
    [`@media only screen and (min-width: ${mediaQueryWidth})`]: style,
  };
};

export const gridStyle = props => {
  const breakpointStyle = breakpoints.reduce((style, breakpoint) => {
    // match props with alignment style and merge for breakpoint
    const alignmentStyle = Object.keys(props).reduce(
      (obj, key) => (props[key] === breakpoint ? { ...obj, ...alignmentStyleMap[key] } : obj),
      {},
    );

    return { ...style, ...wrapStyleWithMediaQuery(breakpoint, alignmentStyle) };
  }, {});

  return {
    boxSizing: 'border-box',
    display: 'flex',
    flex: '0 1 auto',
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginLeft: `-${gutterWidth}`,
    ...breakpointStyle,
  };
};

const calculateGridColumnWidth = index => (100 / 12) * index;

export const gridColumnStyle = props => {
  const breakpointStyle = breakpoints.reduce((style, breakpoint) => {
    const breakpointIndex = props[breakpoint];
    const breakpointOffsetIndex = props[`${breakpoint}Offset`];
    const columnStyle = {
      ...(breakpointIndex
        ? {
            boxSizing: 'border-box',
            flex: `0 0 ${calculateGridColumnWidth(breakpointIndex)}%`,
            maxWidth: `${calculateGridColumnWidth(breakpointIndex)}%`,
            paddingLeft: gutterWidth,
            textAlign: 'left',
          }
        : {}),
      ...(breakpointOffsetIndex
        ? {
            marginLeft: `${calculateGridColumnWidth(breakpointOffsetIndex)}%`,
          }
        : {}),
    };

    return { ...style, ...wrapStyleWithMediaQuery(breakpoint, columnStyle) };
  }, {});

  return {
    boxSizing: 'border-box',
    flex: '1 0 0',
    maxWidth: '100%',
    paddingLeft: gutterWidth,
    textAlign: 'left',
    ...breakpointStyle,
  };
};
