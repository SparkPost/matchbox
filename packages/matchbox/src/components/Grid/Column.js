import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './Grid.module.scss';

class Column extends Component {
  static displayName = 'Grid.Column';

  static propTypes = {
    xs: PropTypes.number,
    sm: PropTypes.number,
    md: PropTypes.number,
    lg: PropTypes.number,
    xl: PropTypes.number,
    xsOffset: PropTypes.number,
    smOffset: PropTypes.number,
    mdOffset: PropTypes.number,
    lgOffset: PropTypes.number,
    children: PropTypes.node
  };

  render() {
    const {
      children,
      xs, sm, md, lg, xl,
      xsOffset, smOffset, mdOffset, lgOffset, xlOffset,
      ...rest
    } = this.props;

    const colClasses = classnames(
      styles.Column,
      xs && styles[`xs-${xs}`],
      sm && styles[`sm-${sm}`],
      md && styles[`md-${md}`],
      lg && styles[`lg-${lg}`],
      xl && styles[`xl-${xl}`],
      xsOffset && styles[`xs-offset-${xsOffset}`],
      smOffset && styles[`sm-offset-${smOffset}`],
      mdOffset && styles[`md-offset-${mdOffset}`],
      lgOffset && styles[`lg-offset-${lgOffset}`],
      xlOffset && styles[`xl-offset-${xlOffset}`]
    );

    return <div className={colClasses} {...rest}>{children}</div>;
  }
}

export default Column;
