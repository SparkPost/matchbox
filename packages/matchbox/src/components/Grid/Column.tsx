import React from 'react';
import styled from 'styled-components';
import { gridColumnStyle } from './styles';

const StyledGridColumn = styled('div')`
  ${gridColumnStyle}
`;

type GridColumnProps = {
  xs?: number;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  xsOffset?: number;
  smOffset?: number;
  mdOffset?: number;
  lgOffset?: number;
  children?: React.ReactNode;
};

function Column({ children, ...props }: GridColumnProps): JSX.Element {
  return <StyledGridColumn {...props}>{children}</StyledGridColumn>;
}

Column.displayName = 'Grid.Column';
export default Column;
