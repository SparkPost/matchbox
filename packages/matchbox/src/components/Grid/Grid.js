import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Column from './Column';
import styles from './Grid.module.scss';

class Grid extends Component {
  static displayName = 'Grid';

  static Column = Column;

  static propTypes = {
    center: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    start: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    end: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    top: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    middle: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    bottom: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    around: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    between: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
    children: PropTypes.node
  };

  render() {
    const {
      children,
      start, center, end,
      top, middle, bottom,
      around, between,
      className
    } = this.props;

    const gridClasses = classnames(
      styles.Grid,
      start && styles[`start-${start}`],
      center && styles[`center-${center}`],
      end && styles[`end-${end}`],
      top && styles[`top-${top}`],
      middle && styles[`middle-${middle}`],
      bottom && styles[`bottom-${bottom}`],
      around && styles[`around-${around}`],
      between && styles[`between-${between}`],
      className
    );

    return <div className={gridClasses}>{children}</div>;
  }
}

export default Grid;
