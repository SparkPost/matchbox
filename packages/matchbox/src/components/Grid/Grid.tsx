import React from 'react';
import styled from 'styled-components';
import { gridStyle } from './styles';
import type { BreakpointsWithoutDefault } from '../../helpers/types';
import Column from './Column';

const StyledGrid = styled('div')<{
  $start?: BreakpointsWithoutDefault;
  $center?: BreakpointsWithoutDefault;
  $end?: BreakpointsWithoutDefault;
  $top?: BreakpointsWithoutDefault;
  $middle?: BreakpointsWithoutDefault;
  $bottom?: BreakpointsWithoutDefault;
  $around?: BreakpointsWithoutDefault;
  $between?: BreakpointsWithoutDefault;
}>`
  ${gridStyle}
`;

type GridProps = {
  start?: BreakpointsWithoutDefault;
  center?: BreakpointsWithoutDefault;
  end?: BreakpointsWithoutDefault;
  top?: BreakpointsWithoutDefault;
  middle?: BreakpointsWithoutDefault;
  bottom?: BreakpointsWithoutDefault;
  around?: BreakpointsWithoutDefault;
  between?: BreakpointsWithoutDefault;
  children: React.ReactNode;
};

function Grid(props: GridProps): JSX.Element {
  const { children, start, center, end, top, middle, bottom, around, between } = props;
  return (
    <StyledGrid
      $start={start}
      $center={center}
      $end={end}
      $top={top}
      $middle={middle}
      $bottom={bottom}
      $around={around}
      $between={between}
    >
      {children}
    </StyledGrid>
  );
}

Grid.displayName = 'Grid';
Grid.Column = Column;
export default Grid;
