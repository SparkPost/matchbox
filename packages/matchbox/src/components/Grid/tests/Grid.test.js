import React from 'react';
import Grid from '../Grid';
import { shallow } from 'enzyme';

describe('Grid', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Grid><p>Grid</p></Grid>);
  });

  it('renders with default props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders other props', () => {
    wrapper.setProps({
      center: 'xs',
      start: 'xs',
      end: 'xs',
      top: 'xs',
      middle: 'xs',
      bottom: 'xs',
      around: 'xs',
      between: 'xs'
    });
    expect(wrapper).toMatchSnapshot();
  });
});
