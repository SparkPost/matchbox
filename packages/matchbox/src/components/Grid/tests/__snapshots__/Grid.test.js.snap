// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders other props 1`] = `
<div
  className="Grid start-xs center-xs end-xs top-xs middle-xs bottom-xs around-xs between-xs"
>
  <p>
    Grid
  </p>
</div>
`;

exports[`Grid renders with default props 1`] = `
<div
  className="Grid"
>
  <p>
    Grid
  </p>
</div>
`;
